text/0095-cognito-construct-library/working-backwards-readme.md

```diff
@@ -0,0 +1,567 @@
+verification](https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-settings-email-phone-verification.html)
+
+*Note*: By default, the email and/or phone number attributes will be marked as required if they are specified as a
+verified contact method. See [attributes](#attributes) section for details about standard attributes.
+
+> These are the defaults that will be part of tsdoc, and not part of the README -
+> * signUp.selfSignUp: true
+> * signUp.tempPasswordValidity: 7 days
+> * signUp.welcomeEmail.subject: 'Thanks for signing up'
+> * signUp.welcomeEmail.body - 'Hello {username}, Your temporary password is {####}'
+> * signUp.welcomeSms.message - 'Your temporary password is {####}'
+> * signUp.verifyContactMethods - Email
+
+> Internal Note: Implemented via UserPool-AdminCreateUserConfig, temp password via UserPool-Policies,
+> verifyContactMethods via UserPool-AutoVerifiedAttributes.
+
+### Sign In
+
+These are the various ways a user of your app can sign in. There are 4 options available with the enum `SignInType`:
+
+* `USERNAME`: Allow signing in using the one time immutable user name that the user chose at the time of sign up.
+* `PREFERRED_USERNAME`: Allow signing in with an alternate user name that the user can change at any time. However, this
+  is not available if the USERNAME option is not chosen.
+* `EMAIL`: Allow signing in using the email address that is associated with the account.
+* `PHONE_NUMBER`: Allow signing in using the phone number that is associated with the account.
+
+> These are the defaults that will be part of tsdoc, and not part of the README -
+> SignInType: `USERNAME`.
+
+Code sample:
+
+```ts
+new UserPool(this, 'myuserpool', {
+  // ...
+  // ...
+  signInType: [ SignInType.USERNAME, SignInType.EMAIL ],
+});
+```
+
+> Internal Note: Implemented via UserPool-UsernameAttributes and -AliasAttributes
+
+### Attributes
+
+These are the set of attributes you want to collect and store with each user in your user pool. Cognito provides a set
+of standard attributes that are available all user pools. Users are allowed to select any of these standard attributes
+to be required. Users will not be able to sign up without specifying the attributes that are marked as required. Besides
+these, additional attributes can be further defined, known as custom attributes.
+
+Learn more on [attributes in Cognito's
+documentation](https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-settings-attributes.html).
+
+The following code sample configures a user pool with two standard attributes (name and address) as required, and adds
+two optional attributes, one as 'string' type and one as 'number' type.
+
+```ts
+new UserPool(this, 'myuserpool', {
+  // ...
+  // ...
+  attributes: {
+    required: [ StandardAttrs.address, StandardAttrs.name ],
+    custom: {
+      'myappid': new StringAttr({ minLen: 5, maxLen: 15 }),
+      'callingcode': new NumberAttr({ min: 1, max: 3 }),
+    },
```

Comment:
Why not an array for custom attributes too?

`custom: [new StringAttr('myappid', {minLen: 5, maxLen: 15 })]`

Reply:
I also don't see from this example code how you make a custom attribute required or a standard attribute optional. The sample seems to conflate required == standard, custom == optional (although that might be a Cognito limitation that I'm unaware of).

Would expect to see this though:

```ts
new UserPool(this, 'myuserpool', {
  // ...
  requiredAttributes: [
    Attribute.NAME,
    Attribute.customString('myappid', { minLen: 5, maxLen: 15 })
  ],
  optionalAttributes: [
    Attribute.ADDRESS,
    Attribute.customNumber('callingcode', { min: 1, max: 3 })
  ]
});
```

Reply:
@rix0rrr -

>  The sample seems to conflate required == standard, custom == optional (although that might be a Cognito limitation that I'm unaware of).

That is exactly Cognito's modeling.

Quoting from the [documentation for standard attributes](https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-settings-attributes.html#cognito-user-pools-standard-attributes) -- "These attributes are available as optional attributes for all users. To make an attribute required, select the check box next to the attribute. ".

Quoting from the [documentation for custom attributes](https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-settings-attributes.html#user-pool-settings-custom-attributes) -- "cannot be required"

Given this information, is my modeling ok?

Reply:
@duarten - switched this from an array to map based on [another comment](https://github.com/aws/aws-cdk-rfcs/pull/91#discussion_r371007799). I believe the map-style is a bit more ergonomic and human friendly.